### PR TITLE
fix(sdkgen/go): Don't set default to zero if env is unset

### LINKED ISSUE #12971: Go provider SDK generation injects empty strings and breaks providers
### What happened?

There is a bug in the provider SDK generator for Go, where it generates code for computing default values for properties in the context of https://www.pulumi.com/docs/concepts/resources/providers/#explicit-provider-configuration and it recently started breaking pulumi-snowlake provider. Note that SDK generator for TypeScript works fine in this case.

Original issue: 
https://github.com/pulumi/pulumi-snowflake/issues/248

### Expected Behavior

Expecting the generated Go provider SDK send the same data to the engine as the generated Node SDK.

### Steps to reproduce

To reproduce, try this Go program:

```go
package main

import (
	"fmt"
	snowflake "github.com/pulumi/pulumi-snowflake/sdk/go/snowflake"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		args := &snowflake.ProviderArgs{
			Account:  pulumi.String("snowhouse"),
			Region:   pulumi.String("us-west-2"),
			Role:     pulumi.String("ENGCLOUD_INGEST_RL"),
			Username: pulumi.String("ENGCLOUD_INGEST_USER"),
			Password: pulumi.String("<REDACTED>"),
		}

		provider, err := snowflake.NewProvider(ctx, "engcloud-ingest-user-provider", args)
		if err != nil {
			return err
		}

		fmt.Println(provider)
		return nil
	})
}
```

This program fails at `pulumi preview` with:

```
    error: rpc error: code = Unknown desc = could not validate provider configuration: 10 errors occurred:
        * Conflicting configuration arguments: "oauth_client_id": conflicts with browser_auth
        * Conflicting configuration arguments: "browser_auth": conflicts with password
        * Conflicting configuration arguments: "private_key_path": conflicts with browser_auth
        * Conflicting configuration arguments: "private_key_passphrase": conflicts with browser_auth
        * Conflicting configuration arguments: "oauth_access_token": conflicts with browser_auth
        * Conflicting configuration arguments: "oauth_refresh_token": conflicts with browser_auth
        * Conflicting configuration arguments: "oauth_redirect_url": conflicts with browser_auth
        * Conflicting configuration arguments: "password": conflicts with browser_auth
        * Conflicting configuration arguments: "oauth_client_secret": conflicts with browser_auth
        * Conflicting configuration arguments: "oauth_endpoint": conflicts with browser_auth
```

Now consider this equivalent TypeScript program, which works as expected without issue:

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as sf from "@pulumi/snowflake";

new sf.Provider("p", {
    account: "snowhouse",
    region: "us-west-2",
    role: "ENGCLOUD_INGREST_RL",
    username: "ENGCLOUD_INGEST_USER",
    password: "<REDACTED>",
})
```

The problem is that Go program sends unwanted empty strings to CheckConfig:

```
{
  "method": "/pulumirpc.ResourceProvider/CheckConfig",
  "request": {
    "urn": "urn:pulumi:dev::snowflake-248::pulumi:providers:snowflake::engcloud-ingest-user-provider",
    "olds": {},
    "news": {
      "account": "snowhouse",
      "browserAuth": false,
      "host": "",
      "oauthAccessToken": "",
      "oauthClientId": "",
      "oauthClientSecret": "",
      "oauthEndpoint": "",
      "oauthRedirectUrl": "",
      "oauthRefreshToken": "",
      "password": "<REDACTED>",
      "port": 0,
      "privateKeyPassphrase": "",
      "privateKeyPath": "",
      "protocol": "",
      "region": "us-west-2",
      "role": "ENGCLOUD_INGEST_RL",
      "username": "ENGCLOUD_INGEST_USER",
      "version": "0.25.0",
      "warehouse": ""
    }
  },
  "response": {},
  "metadata": {
    "kind": "resource",
    "mode": "client",
    "name": "snowflake"
  }
}
```

This can be tracked down to the auto-generated code such as https://github.com/pulumi/pulumi-snowflake/blob/43feda1509fb976f5eae1f1c7264e2a3da5e829a/sdk/go/snowflake/provider.go#L71

Consider code like this:

```go
	if args.OauthClientId == nil {
		args.OauthClientId = pulumi.StringPtr(getEnvOrDefault("", nil, "SNOWFLAKE_OAUTH_CLIENT_ID").(string))
	}
```

If instead it read like this, the problem would go away:

```go
	if args.OauthClientSecret == nil {
		if d := getEnvOrDefault(nil, nil, "SNOWFLAKE_OAUTH_CLIENT_SECRET"); d != nil {
			args.OauthClientSecret = pulumi.StringPtr(d.(string))
		}
	}
```

### Output of `pulumi about`

```
CLI          
Version      3.67.0
Go Version   go1.20.4
Go Compiler  gc

Plugins
NAME       VERSION
go         unknown
snowflake  0.25.0

Host     
OS       darwin
Version  13.3.1
Arch     x86_64

This project is written in go: executable='/Users/t0yv0/.nix-profile/bin/go' version='go version go1.20.1 darwin/amd64'

Current Stack: t0yv0/snowflake-248/dev

TYPE                        URN
pulumi:pulumi:Stack         urn:pulumi:dev::snowflake-248::pulumi:pulumi:Stack::snowflake-248-dev
pulumi:providers:snowflake  urn:pulumi:dev::snowflake-248::pulumi:providers:snowflake::engcloud-ingest-user-provider


Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/t0yv0
User           t0yv0
Organizations  t0yv0, pulumi

Dependencies:
NAME                                    VERSION
github.com/pulumi/pulumi-snowflake/sdk  0.25.0
github.com/pulumi/pulumi/sdk/v3         3.60.1

Pulumi locates its logs in /var/folders/gk/cchgxh512m72f_dmkcc3d09h0000gp/T/ by default
```

### Additional context

Tied to a customer request, but workarounds may be possible at provider/bridge level, looking into that. 

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by abhinav**
Each commit in this change is separately reviewable.

---

sdkgen has a discrepancy in behavior for default values.
There are three scenarios:

- default value
- default value and an environment variable fallback
- only an environment variable fallback

We handle the first two correctly:
the value is set only if there was something to set it to.
But in the third case, we set variables to their zero value
when the environment variable is empty.

    // Current implementation effectively does:

    if v := os.Getenv("whatever"); v != "" {
        out.Field = pointerOf(v)
    } else {
        out.Field = pointerOf("")
    }

To fix this, we generate code to set these values
only if the environment variable is set.

Testing:
Includes a regression test.

Implementation notes:
Previosly, getDefaultValue generated an expression
holding the default value.
For fixed default values, it was just the value.
For default values with environment variable fallback,
it took the form:

    getEnvOrDefault($default, $parser, $env1, $env2, ...).($type)

For example:

    getEnvOrDefault("", nil, "ITEM_NAME").(string)

Given that getDefaultValue returned an expression,
its consumers were able to do something like:

    if result == nil {
        result = getEnvOrDefault(...).(string)
    }

As part of this fix, we need to stop assinging to result
if the environment variable is unset,
so we want getDefaultValue to generate:

    if result == nil {
        if d := getEnvOrDefault(...); d != nil {
            result = d
        }
    }

To accomplish this, we switch getDefaultValue to setDefaultValue.
Instead of returning an expression, it accepts a callback.
It generates the blocks it needs to, and invokes the callback
with an expression holding the default value if necessary.

Resolves #12971

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12976** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12976" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/12976?utm_source=stack-comment).

File: pkg/codegen/testing/test/testdata/regress-go-12971/go/example/pulumiTypes.go [hunk 1]
```diff
@@ -23,16 +23,25 @@ func (val *World) Defaults() *World {
...
+
+		}
 	}
 	if tmp.RadiusKm == nil {
-		radiusKm_ := getEnvOrDefault(0.0, parseEnvFloat, "WORLD_RADIUS_KM").(float64)
-		tmp.RadiusKm = &radiusKm_
+		if d := getEnvOrDefault(nil, parseEnvFloat, "WORLD_RADIUS_KM"); d != nil {
+			radiusKm_ := d.(float64)
+			tmp.RadiusKm = &radiusKm_
+
```
Comment (t0yv0) [inline pkg/codegen/testing/test/testdata/regress-go-12971/go/example/pulumiTypes.go:43]: Nit: remove empty line here?

**Comment by RobbieMcKinstry** [comment]
@abhinav I'm struggling a bit to understand where and when this change is happening. I think my mental model is v wrong.

> sdkgen has a discrepancy in behavior for default values. There are three scenarios:
>    default value
>    default value and an environment variable fallback
>    only an environment variable fallback

Are these env vars read during code-gen time? Or are these values like `GITHUB_TOKEN` which are read when the user's program runs?

How am I to interpret the case where the value is unassigned? This means that the variable won't be in scope at all if the Environment Variable isn't present (at codegen time)? That would make sense to me.

**Comment by RobbieMcKinstry** [comment]
I think I get it: the values set to `CheckConfig` aren't necessarily fixed. It's not variable assignments that need to be in scope, like e.g. a variable in Go needs to be in scope if its' used, and used if its in scope. 

Does that sound like the right mental model? I went back and read Anton's issue which I think helped me.

**Review by RobbieMcKinstry — COMMENTED** [review]
Left a question. No issues with the code; the comments are very helpful.

File: pkg/codegen/testing/test/testdata/regress-go-12971/schema.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "world",
+  "version": "0.0.1",
+  "types": {
+    "world::World": {
+      "type": "object",
+      "properties": {
+        "name": {"type": "string", "defaultInfo": {"environment": ["WORLD_NAME"]}},
+        "radiusKm": {"type": "number", "defaultInfo": {"environment": ["WORLD_RADIUS_KM"]}},
+        "populated": {"type": "boolean", "defaultInfo": {"environment": ["WORLD_POPULATED"]}}
```
Comment (justinvp) [inline pkg/codegen/testing/test/testdata/regress-go-12971/schema.json:10]: Should we also include a `"provider"` resource with similar input properties to these, to ensure the generated  `NewProvider` func is what we expect (which is where the problem is in this case for the Snowflake provider)? Right now, we're only testing the generated `Default` method for types.

Something like this in the schema:

```json
"types": {
    ...
},
"provider": {
    "inputProperties": {
        "name": {"type": "string", "defaultInfo": {"environment": ["WORLD_NAME"]}},
        "radiusKm": {"type": "number", "defaultInfo": {"environment": ["WORLD_RADIUS_KM"]}},
        "populated": {"type": "boolean", "defaultInfo": {"environment": ["WORLD_POPULATED"]}}
    }
}
```

Reference: https://github.com/pulumi/pulumi-snowflake/blob/0742790b87a393e48b7b2483c7cb0837cd3bc4b5/provider/cmd/pulumi-resource-snowflake/schema.json#L1993-L2002

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: pkg/codegen/testing/test/testdata/regress-go-12971/schema.json (hunk 2 above)
Comment (abhinav) [reply pkg/codegen/testing/test/testdata/regress-go-12971/schema.json:10]: Good call. I'm changing the test case to verify provider as well as config.
That should cover all cases.

**Review by justinvp — APPROVED** [review]

**Comment by t0yv0** [comment]
@RobbieMcKinstry can I answer any questions further? Eager to try this out as P1 remedy once on master. What Abhinav says is right we used to send `{"foo": ""}` to CheckConfig/Configure instead of `{}` (no foo) in the default case (empty environment), which tripped it up. This change feels good the new code is more natural.

Separately I want to follow through myself why does Snowflake provider, being a bridged provider, even expose these ENV vars to codegen. I thought that bridged providers do it's own hocus-pocus with default values and they like to control that to for example not set a default from an env var if it conflicts with another field, in fact the bridge reads ENV variables itself, so it's a bit strange it also asks SDKs to do so.

**Comment by abhinav** [comment]
@RobbieMcKinstry I believe the comments above were not blocking, so I'm landing. (LMK if I misinterpreted.)

**Comment by abhinav** [comment]
bors merge

**Comment by RobbieMcKinstry** [comment]
They were indeed non-blocking! :D I didn't want to slap a green check on it if I didn't understand the PR, but don't let my questions hold up a fix for a P1! :) 
 
Just FYI for the future, if I leave blocking comments, I will mark the review as "changes requested" and explicitly say the comments are blocking in a message! :D 

Thanks for the explanation! That helped me understand! I was originally confused whether `x` would be in scope in the case where nothing was provided (no default or env var), but I see now that it's in scope either way; it's just that the value will be set differently in each of the cases.

Much appreciated ya'll!

**Pull request merged**